publish-externals/src/index.ts

```diff
@@ -2,35 +2,114 @@ import path from "path"
 import * as hashnode from "./utils/hashnode"
 import {extractFrontMatterAndContent} from "./utils/markdown"
 import * as devTo from "./utils/devto"
+import snapshot from "../snapshot.json"
+import fs from "fs"
+import {createHash} from "crypto"
 
-const ExternalPublications = [
+interface Blog {
+  file: string
+  platforms: {
+    [platform: string]: {
+      published: boolean
+      lastUpdatePublished: boolean
+      lastSuccessfulPublishedAt?: string
+    }
+  }
+  hash?: string
+}
+
+interface ExternalPublication {
+  name: string
+  handler: (frontMatter: FrontMatter, content: string) => Promise<void>
+}
+
+interface FrontMatter {
+  slug: string
+}
+
+const ExternalPublications: ExternalPublication[] = [
   {name: "Hashnode", handler: hashnode.handler},
   {name: "Dev.to", handler: devTo.handler},
 ]
 
 const main = async () => {
-  const changedFiles = process.argv[2].split(" ")
-  const errors = []
-  for (const file of changedFiles) {
-    if (file.startsWith("blog/")) {
-      const filePath = path.join(__dirname, "../../", file)
-      const {frontMatter, content} = extractFrontMatterAndContent(filePath)
-      for (let publication of ExternalPublications) {
-        console.log(`[${publication.name}] ${frontMatter.slug} ... publishing ⏳`)
-        try {
-          await publication.handler(frontMatter, content)
-          console.log(`[${publication.name}] Success ${frontMatter.slug} ... succeeded ✅`)
-        } catch (error) {
-          errors.push(error)
-          console.error(`[${publication.name}] Failure ${frontMatter.slug} ... failed 💀`)
+  const addedFilesArg = process.argv[2]
+  const addedFiles = addedFilesArg ? addedFilesArg.split(" ") : []
+  const blogs: {[key: string]: Blog} = snapshot.blogs || {}
+
+  if (addedFiles.length === 0 && Object.keys(blogs).length === 0) {
+    console.log("No changes detected. Exiting.")
+    return
+  }
+
+  try {
+    for (let publication of ExternalPublications) {
+      for (const file of addedFiles) {
+        if (file.startsWith("blog/")) {
+          await publish(file, blogs, publication)
+        }
+      }
+
+      for (const slug in blogs) {
+        const file = blogs[slug].file
+        if (!addedFiles.includes(file)) {
+          const {content} = extractFrontMatterAndContent(path.join(__dirname, "../../", file))
+          const contentHash = createHash("sha256").update(content).digest("hex")
+
+          if (!blogs[slug].hash || blogs[slug].hash !== contentHash) {
+            await publish(file, blogs, publication)
```

Comment:
Take a CLI argument `--publish`. If passed only then should you actually publish, otherwise we should simply perform a dry run. 
In the dry run mode, we should "fake" meaning just say what's going to be published instead of actually publishing.